libavfilter/vf_deinterlace_qsv.c

```diff
@@ -115,441 +70,54 @@ static int qsvdeint_query_formats(AVFilterContext *ctx)
-        par.vpp.In.FrameRateExtD = ctx->inputs[0]->frame_rate.den;
-    } else {
-        par.vpp.In.FrameRateExtN = ctx->inputs[0]->time_base.num;
-        par.vpp.In.FrameRateExtD = ctx->inputs[0]->time_base.den;
-    }
-
-    par.vpp.Out = par.vpp.In;
-
-    if (ctx->outputs[0]->frame_rate.num) {
-        par.vpp.Out.FrameRateExtN = ctx->outputs[0]->frame_rate.num;
-        par.vpp.Out.FrameRateExtD = ctx->outputs[0]->frame_rate.den;
-    } else {
-        par.vpp.Out.FrameRateExtN = ctx->outputs[0]->time_base.num;
-        par.vpp.Out.FrameRateExtD = ctx->outputs[0]->time_base.den;
-    }
-
-    /* Print input memory mode */
-    ff_qsvvpp_print_iopattern(ctx, par.IOPattern & 0x0F, "VPP");
-    /* Print output memory mode */
-    ff_qsvvpp_print_iopattern(ctx, par.IOPattern & 0xF0, "VPP");
-    err = MFXVideoVPP_Init(s->session, &par);
-    if (err < 0)
-        return ff_qsvvpp_print_error(ctx, err,
-                                     "Error opening the VPP for deinterlacing");
-    else if (err > 0) {
-        ff_qsvvpp_print_warning(ctx, err,
-                                "Warning in VPP initialization");
-        return AVERROR_UNKNOWN;
-    }
-
-    return 0;
-}
-
 static int qsvdeint_config_props(AVFilterLink *outlink)
 {
     AVFilterContext *ctx = outlink->src;
     AVFilterLink *inlink = ctx->inputs[0];
-    QSVDeintContext  *s = ctx->priv;
-    int ret;
+    QSVDeintContext   *s = ctx->priv;
+    QSVVPPParam    param = { NULL };
+    mfxExtBuffer *ext_buf[1];
+    enum AVPixelFormat in_format;
 
     qsvdeint_uninit(ctx);
 
-    s->last_pts = AV_NOPTS_VALUE;
+    outlink->w          = inlink->w;
+    outlink->h          = inlink->h;
     outlink->frame_rate = av_mul_q(inlink->frame_rate,
                                    (AVRational){ 2, 1 });
     outlink->time_base  = av_mul_q(inlink->time_base,
                                    (AVRational){ 1, 2 });
 
-    /* check that we have a hw context */
-    if (!inlink->hw_frames_ctx) {
-        av_log(ctx, AV_LOG_ERROR, "No hw context provided on input\n");
-        return AVERROR(EINVAL);
-    }
-
-    s->hw_frames_ctx = av_buffer_ref(inlink->hw_frames_ctx);
-    if (!s->hw_frames_ctx)
-        return AVERROR(ENOMEM);
-
-    av_buffer_unref(&outlink->hw_frames_ctx);
-    outlink->hw_frames_ctx = av_buffer_ref(inlink->hw_frames_ctx);
-    if (!outlink->hw_frames_ctx) {
-        qsvdeint_uninit(ctx);
-        return AVERROR(ENOMEM);
-    }
-
-    ret = init_out_session(ctx);
-    if (ret < 0)
-        return ret;
-
-
-    return 0;
-}
-
-static void clear_unused_frames(QSVDeintContext *s)
-{
-    QSVFrame *cur = s->work_frames;
-    while (cur) {
-        if (!cur->surface.Data.Locked) {
-            av_frame_free(&cur->frame);
-            cur->queued = 0;
-        }
-        cur = cur->next;
-    }
-}
-
-static int get_free_frame(QSVDeintContext *s, QSVFrame **f)
-{
-    QSVFrame *frame, **last;
-
-    clear_unused_frames(s);
+    if (inlink->format == AV_PIX_FMT_QSV) {
```

Comment:
could we have a function for this logical?

Reply:
This is not the final version. I will re-use all functions in vf_vpp_qsv.c for deinterlace_qsv filter too.